File: asr/harmony/sdk/src/main/ets/com/amphion/asr/SpeakerVadScoreScheduler.ts
/**
 * Anchors Speaker VAD score deadlines to absolute PCM sample positions in one native segment.
 *
 * The runtime stops each internal feed at samplesUntilNextScore(), then calls observe(). This keeps
 * the score timeline independent of how public writeAudio calls partition the same PCM.
 */
export class SpeakerVadScoreScheduler {
  readonly windowSamples: number;
  readonly hopSamples: number;
  private consumedSamples: number = 0;
  private firstScoreSample: number;
  private nextScoreSample: number;

  constructor(windowSamples: number, hopSamples: number) {
    this.windowSamples = Math.round(windowSamples);
    this.hopSamples = Math.round(hopSamples);
    if (this.windowSamples <= 0) throw new Error('windowSamples must be > 0');
    if (this.hopSamples <= 0) throw new Error('hopSamples must be > 0');
    this.firstScoreSample = Math.max(this.windowSamples, this.hopSamples);
    this.nextScoreSample = this.firstScoreSample;
  }

  totalSamples(): number {
    return this.consumedSamples;
  }

  samplesUntilNextScore(): number {
    return Math.max(1, this.nextScoreSample - this.consumedSamples);
  }

  /** Returns true exactly when the accepted slice ends at a score deadline. */
  observe(samples: number): boolean {
    const accepted = Math.round(samples);
    if (accepted < 0) throw new Error('samples must be >= 0');
    if (accepted > this.samplesUntilNextScore()) {
      throw new Error('accepted slice crosses Speaker VAD score deadline');
    }
    this.consumedSamples += accepted;
    if (this.consumedSamples !== this.nextScoreSample) return false;

    if (this.nextScoreSample === this.firstScoreSample) {
      const remainder = this.firstScoreSample % this.hopSamples;
      this.nextScoreSample += remainder === 0 ? this.hopSamples : this.hopSamples - remainder;
    } else {
      this.nextScoreSample += this.hopSamples;
    }
    return true;
  }

  reset(): void {
    this.consumedSamples = 0;
    this.nextScoreSample = this.firstScoreSample;
  }
}

File: asr/android/sdk/src/main/java/com/amphion/asr/internal/SpeakerVadScoreScheduler.kt
package com.amphion.asr.internal

/**
 * Anchors Speaker VAD score deadlines to absolute PCM sample positions in one native segment.
 *
 * Callers must stop each internal feed at [samplesUntilNextScore], then call [observe]. This keeps
 * the score timeline independent of how the public writeAudio calls partition the same PCM.
 */
internal class SpeakerVadScoreScheduler(
    val windowSamples: Int,
    val hopSamples: Int,
) {
    init {
        require(windowSamples > 0) { "windowSamples must be > 0" }
        require(hopSamples > 0) { "hopSamples must be > 0" }
    }

    var totalSamples: Int = 0
        private set

    private val firstScoreSample: Int = maxOf(windowSamples, hopSamples)
    private var nextScoreSample: Int = firstScoreSample

    fun samplesUntilNextScore(): Int = (nextScoreSample - totalSamples).coerceAtLeast(1)

    /** Returns true exactly when the accepted slice ends at a score deadline. */
    fun observe(samples: Int): Boolean {
        require(samples >= 0) { "samples must be >= 0" }
        require(samples <= samplesUntilNextScore()) {
            "accepted slice crosses Speaker VAD score deadline"
        }
        totalSamples += samples
        if (totalSamples != nextScoreSample) return false

        nextScoreSample += if (nextScoreSample == firstScoreSample) {
            val remainder = firstScoreSample % hopSamples
            if (remainder == 0) hopSamples else hopSamples - remainder
        } else {
            hopSamples
        }
        return true
    }

    fun reset() {
        totalSamples = 0
        nextScoreSample = firstScoreSample
    }
}
